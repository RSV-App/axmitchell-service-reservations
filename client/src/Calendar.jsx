import React from 'react';
import moment from 'moment';
import dateFns from 'date-fns';
import PropTypes from 'prop-types';

import '../../public/styles.css';

class Calendar extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      weekDays: moment.weekdaysShort(),
      currentMonth: new Date(),
      currentDay: new Date(),
      clickedDay: new Date(),
    };

    this.prevMonthClick = this.prevMonthClick.bind(this);
    this.nextMonthClick = this.nextMonthClick.bind(this);
    this.changeDayClick = this.changeDayClick.bind(this);
    this.setWrapperRef = this.setWrapperRef.bind(this);
    this.handleClickOutside = this.handleClickOutside.bind(this);
  }

  componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside);
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside);
  }

  setWrapperRef(node) {
    this.wrapperRef = node;
  }

  handleClickOutside(event) {
    if (this.wrapperRef && !this.wrapperRef.contains(event.target)) {
      this.props.hideCalendar();
    }
  }

  currentDay() {
    return dateFns.format(this.state.currentDay, 'D');
  }

  changeDayClick(e) {
    if (dateFns.compareAsc(e, dateFns.subDays(new Date(), 1)) === 1) {
      this.props.handleDate(e);
      this.props.changeDate(e);
      this.props.getBtnBack();
      this.props.hideCalendar();
    }
  }

  prevMonthClick() {
    const todayDate = moment().format('MMMM YYYY');
    if (dateFns.format(this.state.currentMonth, 'MMMM YYYY') !== todayDate) {
      this.setState({
        currentMonth: dateFns.subMonths(this.state.currentMonth, 1),
      });
    }
  }

  nextMonthClick() {
    this.setState({
      currentMonth: dateFns.addMonths(this.state.currentMonth, 1),
    });
  }

  renderDays() {
    const dateFormat = 'dddd';
    const days = [];

    const startDate = dateFns.startOfWeek(this.state.currentMonth);

    for (let i = 0; i < 7; i++) {
      days.push(
        <td key={i} className="calendar-day-res">{dateFns.format(dateFns.addDays(startDate, i), dateFormat)}</td>,
      );
    }
  }

  renderCells() {
    const monthStart = dateFns.startOfMonth(this.state.currentMonth);
    const startDate = dateFns.startOfWeek(monthStart);

    const dateFormat = 'D';
    const rows = [];
    let rowCount = 0;
    let days = [];
    let day = startDate;
    let formattedDate = '';

    while (rowCount < 6) {
      for (let i = 0; i < 7; i++) {
        formattedDate = dateFns.format(day, dateFormat);
        const cloneDay = day;
        let currentDay = '';

        // conditional styling for dates older than today's date
        let pastCurrentDay = '';
        if (dateFns.compareAsc(day, dateFns.subDays(new Date(), 1)) === -1) {
          pastCurrentDay = 'pastCurrentDay';
        }

        // conditional styling for prev month dates
        let prevMonthStyle = '';
        const getPrevMonth = dateFns.subMonths(this.state.currentMonth, 1);
        const prevMonth = dateFns.format(getPrevMonth, 'MMMM');

        if (prevMonth === dateFns.format(day, 'MMMM')) {
          prevMonthStyle = 'prevMonthStyle';
        }

        // conditional styling for next month dates
        let nextMonthStyle = '';
        const getNextMonth = dateFns.addMonths(this.state.currentMonth, 1);
        const nextMonth = dateFns.format(getNextMonth, 'MMMM');

        if (nextMonth === dateFns.format(day, 'MMMM')) {
          nextMonthStyle = 'nextMonthStyle';
        }

        // hover date styling
        let hoverDates = '';
        if (pastCurrentDay) {
          prevMonthStyle = '';
        } else {
          hoverDates = 'hoverStyle';
        }

        if (!pastCurrentDay && !prevMonthStyle && !nextMonthStyle) {
          currentDay = dateFns.format(day, 'M/D/YY') === dateFns.format(this.props.selectedDate, 'M/D/YY') ? 'selectedDate' : '';
        }

        if (currentDay === 'selectedDate') {
          hoverDates = '';
        }

        const classes = `${prevMonthStyle} ${nextMonthStyle} ${pastCurrentDay} ${hoverDates} calendar-day-res ${currentDay}`;

        days.push(
          <td key={i} onClick={() => this.changeDayClick(dateFns.parse(cloneDay))} className={classes}>
            {formattedDate}
          </td >,
        );

        day = dateFns.addDays(day, 1);
      }

      rows.push(
        <tr key={day}>{days}</tr>,
      );

      rowCount++;
      days = [];
    }

    return rows;
  }

  render() {
    const weekDaysName = this.state.weekDays.map(day => <th key={day} className="week-day-res">{day}</th>);

    // conditional styling for prev button
    let prevMonthButton = 'prev-month';
    const currentMonth = dateFns.format(new Date(), 'MMMM YYYY');
    const clickedMonth = dateFns.format(this.state.currentMonth, 'MMMM YYYY');
    if (currentMonth === clickedMonth) {
      prevMonthButton = 'greyed-prev-month';
    }

    return (
      <div onClick={this.childClickHandler} className="res-calendar-wrapper" ref={this.setWrapperRef}  >

        <div className="res-month-title">
          <div className="col-start">
            <div className={prevMonthButton} onClick={this.prevMonthClick}></div>
          </div>

          <div className="col col-center">
            <span className="res-calendar-nav">{dateFns.format(this.state.currentMonth, 'MMMM YYYY')}</span>
          </div>

          <div className="col-end">
            <div className="next-month" onClick={this.nextMonthClick}></div>
          </div>
        </div>

        <table className="reservations-calendar">
          <thead>
            <tr>{weekDaysName}</tr>
          </thead>
          <tbody>
            {this.renderDays()}
            {this.renderCells()}
          </tbody>
        </table>
      </div>
    );
  }
}

Calendar.propTypes = {
  handleDate: PropTypes.func,
  hideCalendar: PropTypes.func,
  changeDate: PropTypes.func,
  getBtnBack: PropTypes.func,
  selectedDate: PropTypes.any.isRequired,
};

export default Calendar;
