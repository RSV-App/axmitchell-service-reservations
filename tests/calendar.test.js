import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

import Calendar from '../client/src/Calendar.jsx';

configure({ adapter: new Adapter() });

it('renders correctly', () => {
  const tree = renderer
    .create(<Calendar />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});


describe('Calendar Rendering', () => {
  it('renders a table for calendar dates', () => {
    const wrapper = shallow(<Calendar />);
    const cells = wrapper.find('.calendar-day-res');
    const headers = wrapper.find('.week-day-res');
    expect(wrapper.find('.reservations-calendar').length).toBe(1);
    expect(cells.length).toBe(41);
    expect(headers.length).toBe(7);
  });
});
