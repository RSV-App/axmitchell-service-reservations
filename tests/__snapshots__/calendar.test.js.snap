// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="res-calendar-wrapper"
>
  <div
    className="res-month-title"
  >
    <div
      className="col-start"
    >
      <div
        className="greyed-prev-month"
        onClick={[Function]}
      />
    </div>
    <div
      className="col col-center"
    >
      <span
        className="res-calendar-nav"
      >
        May 2019
      </span>
    </div>
    <div
      className="col-end"
    >
      <div
        className="next-month"
        onClick={[Function]}
      />
    </div>
  </div>
  <table
    className="reservations-calendar"
  >
    <thead>
      <tr>
        <th
          className="week-day-res"
        >
          Sun
        </th>
        <th
          className="week-day-res"
        >
          Mon
        </th>
        <th
          className="week-day-res"
        >
          Tue
        </th>
        <th
          className="week-day-res"
        >
          Wed
        </th>
        <th
          className="week-day-res"
        >
          Thu
        </th>
        <th
          className="week-day-res"
        >
          Fri
        </th>
        <th
          className="week-day-res"
        >
          Sat
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="prevMonthStyle  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          28
        </td>
        <td
          className="prevMonthStyle  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          29
        </td>
        <td
          className="prevMonthStyle  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          30
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          1
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          2
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          3
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          4
        </td>
      </tr>
      <tr>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          5
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          6
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          7
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          8
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          9
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          10
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          11
        </td>
      </tr>
      <tr>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          12
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          13
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          14
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          15
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          16
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          17
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          18
        </td>
      </tr>
      <tr>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          19
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          20
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          21
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          22
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          23
        </td>
        <td
          className="  pastCurrentDay calendar-day-res"
          onClick={[Function]}
        >
          24
        </td>
        <td
          className="   calendar-day-restoday"
          onClick={[Function]}
        >
          25
        </td>
      </tr>
      <tr>
        <td
          className="   calendar-day-res"
          onClick={[Function]}
        >
          26
        </td>
        <td
          className="   calendar-day-res"
          onClick={[Function]}
        >
          27
        </td>
        <td
          className="   calendar-day-res"
          onClick={[Function]}
        >
          28
        </td>
        <td
          className="   calendar-day-res"
          onClick={[Function]}
        >
          29
        </td>
        <td
          className="   calendar-day-res"
          onClick={[Function]}
        >
          30
        </td>
        <td
          className="   calendar-day-res"
          onClick={[Function]}
        >
          31
        </td>
        <td
          className=" nextMonthStyle  calendar-day-res"
          onClick={[Function]}
        >
          1
        </td>
      </tr>
      <tr>
        <td
          className=" nextMonthStyle  calendar-day-res"
          onClick={[Function]}
        >
          2
        </td>
        <td
          className=" nextMonthStyle  calendar-day-res"
          onClick={[Function]}
        >
          3
        </td>
        <td
          className=" nextMonthStyle  calendar-day-res"
          onClick={[Function]}
        >
          4
        </td>
        <td
          className=" nextMonthStyle  calendar-day-res"
          onClick={[Function]}
        >
          5
        </td>
        <td
          className=" nextMonthStyle  calendar-day-res"
          onClick={[Function]}
        >
          6
        </td>
        <td
          className=" nextMonthStyle  calendar-day-res"
          onClick={[Function]}
        >
          7
        </td>
        <td
          className=" nextMonthStyle  calendar-day-res"
          onClick={[Function]}
        >
          8
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
